daemon/cmd/daemon.go

```diff
@@ -1159,6 +1168,43 @@ func (d *Daemon) bootstrapClusterMesh(nodeMngr *nodemanager.Manager) {
 	bootstrapStats.clusterMeshInit.End(true)
 }
 
+// ReloadOnDeviceChange regenerates device related information and reloads the datapath.
+func (d *Daemon) ReloadOnDeviceChange(devices []string) {
+	log.WithField(logfields.Devices, devices).Infof("Datapath reload initiated due to changed devices.")
+
+	option.Config.Devices = devices
+
+	if option.Config.EnableIPv4Masquerade && option.Config.EnableBPFMasquerade {
+		if err := node.InitBPFMasqueradeAddrs(devices); err != nil {
+			log.Warnf("InitBPFMasqueradeAddrs failed: %s", err)
+		}
+	}
+
+	if option.Config.EnableNodePort {
+		if err := node.InitNodePortAddrs(devices, option.Config.LBDevInheritIPAddr); err != nil {
+			log.WithError(err).Warn("Failed to initialize NodePort addresses")
+		} else {
+			// Synchronize services and endpoints to reflect new addresses onto lbmap.
+			d.svc.SyncServicesOnDeviceChange(d.Datapath().LocalNodeAddressing())
+			d.syncEndpointsAndHostIPs()
+		}
+	}
+
+	// Recreate node_config.h to reflect the mac addresses of the new devices.
+	if err := d.createNodeConfigHeaderfile(); err != nil {
```

Comment:
What guarantees do we have about atomicity here?

Endpoints could be currently regenerating, plausibly reading this file at the same time as we do the rewrite..?

Reply:
Good catch. The `InitBPFMasqueradeAddrs` and `InitNodePortAddrs` are also unsafe since they mutate maps that aren't protected as the assumption in the agent is that they're static... and probably `syncEndpointsAndHostIPs` is also unsafe too since that could be triggered from a controller concurrently with this.

I can fix the node config header issues by taking the compilation lock. The other issues related to mutating  globals (option.Config.Devices, node.ipv4NodePortAddrs, etc.) will require bigger refactoring. Feels a bit like all things like this that may change over time would need to be event-based to make sure there's reaction to the changes.

What do you think, should I start looking into making these mutations safer or would we be ok with putting this in behind an experimental flag and then fixing the potential races in a follow up?

Reply:
I guess the 'simple' answer is to introduce some exclusion locks in the various places so that we at least protect these critical sections. Whether extending the critical section of the compilation lock, or adding individual locks for portions of this reconfiguration, or atomics. If there's a low-effort answer with these, then I think it would be worthwhile. Without that, I will have trouble suggesting anyone to turn the feature on. 

Out of curiousity, do we have a way to manually trigger the go race checker? Does it identify issues at this point? I believe we have a CI job for this which doesn't run on every PR, so it'd be useful to know whether this PR would introduce issues there; once you make a job like that break, it's difficult to get it back green again. Especially if we're assuming that more invasive work is required to address the issue.

Reply:
yes, `/test-race`: https://docs.cilium.io/en/stable/contributing/testing/ci/#testing-with-race-condition-detection-enabled